Possibility of CPU load-limiting (multi-process) tasks

sawenzel opened the pull request.

sawenzel committed:
Include cpulimit tool

Include cpulimit tool from commit f4d2682804931
https://github.com/opsengine/cpulimit for easy
distribution/use within O2.

sawenzel committed:
Improvements to taskwrapper

* More direct CPU monitoring:
  - calc CPU usage directly in script

* Prototypic possibility to limit the total CPU load
  of a task running within taskwrapper by using
  cpulimit (adaptively).

  This may be useful for several reasons. Here, it started
  as an investigation to prevent CPU spikes from workflows
  when running on single-core queues (until DPL has it's own
  way of forcing single-core). Or one might just want to emulate
  the constraints of a GRID multicore queue locally.

  By default the feature is switched off.

  It can be enabled via environment variables. For example

  ```
  JOBUTILS_LIMITLOAD=300 taskwrapper task1.log task1
  ```

  will constrain task1 to using ~3 full CPU cores.

  The regulation mechanism is quite simplistic but should be enough
  to demonstrate the feature.

davidrohr commented:
I like that the impact on the jobutils script is minimum, I don't see any downside, so this is fine with me.
Just 2 comments:
- I don't really like to have much external code in O2, perhaps we should create an O2_externals package to collect such stuff?
- I guess also the codechecker will pick up this code and throw some errors, won't it?

sawenzel committed:
Fix compiler warning on MacOS

sawenzel commented:
@davidrohr : I agree with you that a collection of simple externals could be made and would be preferential. But then the question is whether we need the tools during build or runtime etc. For the moment, I would just keep it simple for a bit of testing. Clang-format should be fine and codechecker might complain (not sure -- usually it only takes stuff within o2 namespace) once.

aphecetche committed:
Allow some executables to not follow the convention.

That list should be kept extremely short, of course.

Also amended the list of known subsystems to reduce the number of
(harmless but annoying) warnings.
